libz-rs-sys/src/gz.rs

```diff
@@ -1296,6 +1303,340 @@ unsafe fn gz_decomp(state: &mut GzState) -> Result<(), ()> {
+    }
+
+    Ok(())
+}
+
+// Compress whatever is at avail_in and next_in (unless in direct mode) and write
+// to the output file.
+//
+// # Returns
+//
+// - `Ok` on success.
+// - `Err` on error.
+fn gz_comp(state: &mut GzState, flush: c_int) -> Result<(), ()> {
+    // Allocate memory if this is the first time through
+    if state.input.is_null() && gz_init(state).is_err() {
+        return Err(());
+    }
+
+    // Write directly if requested.
+    if state.direct {
+        let got = unsafe {
+            libc::write(
+                state.fd,
+                state.stream.next_in.cast::<c_void>(),
+                state.stream.avail_in as _,
+            )
+        };
+        if got < 0 || got as c_uint != state.stream.avail_in {
+            unsafe { gz_error(state, Some((Z_ERRNO, "write error"))) };
+            return Err(());
+        }
+        state.stream.avail_in = 0;
+        return Ok(());
+    }
+
+    // Check for a pending reset.
+    if state.reset {
+        // Don't start a new gzip stream unless there is data to write.
+        if state.stream.avail_in == 0 {
+            return Ok(());
+        }
+        // Safety: `state.reset` is set only in `gz_comp`, which first initializes
+        // `state.stream` using `deflateInit2_`.
+        let _ = unsafe { deflateReset(&mut state.stream) };
+        state.reset = false;
+    }
+
+    // Run deflate on the provided input until it produces no more output.
+    let mut ret = Z_OK;
+    loop {
+        // Write out current buffer contents if full, or if flushing, but if
+        // doing Z_FINISH then don't write until we get to Z_STREAM_END.
+        if state.stream.avail_out == 0
+            || (flush != Z_NO_FLUSH && (flush != Z_FINISH || ret == Z_STREAM_END))
+        {
+            // Safety: Within this gz module, `state.stream.next` and `state.stream.next_out`
+            // always point within the same allocated object, `state.stream.output`.
+            let have = unsafe { state.stream.next_out.offset_from(state.next) };
+            if have != 0 {
+                let ret = unsafe { libc::write(state.fd, state.next.cast::<c_void>(), have as _) };
+                if ret != have as _ {
+                    unsafe { gz_error(state, Some((Z_ERRNO, "write error"))) };
+                    return Err(());
+                }
+            }
+            if state.stream.avail_out == 0 {
+                state.stream.avail_out = state.out_size as _;
+                state.stream.next_out = state.output;
+            }
+            state.next = state.stream.next_out;
+        }
+
+        // Compress.
+        let mut have = state.stream.avail_out;
+        ret = unsafe { deflate(&mut state.stream, flush) };
+        if ret == Z_STREAM_ERROR {
+            unsafe {
+                gz_error(
+                    state,
+                    Some((Z_STREAM_ERROR, "internal error: deflate stream corrupt")),
+                )
+            };
+            return Err(());
```

Comment:
```suggestion
            let msg = "internal error: deflate stream corrupt";
            unsafe { gz_error(state, Some((Z_STREAM_ERROR, msg))) };
            return Err(());
```

These `gz_error` calls are taking up a lot of lines, so something like the above is totally OK to reduce the impact a bit.